### PR TITLE
provider/heroku: Fix heroku_cert update of ssl cert

### PR DESCRIPTION
**Pull request by wchrisjohnson**
This change fixes https://github.com/hashicorp/terraform/issues/5930
Also:
- slight refactor of the update method for the sake of logging.
- beefed up tests around SSL endpoint
- broke out tests to differentiate EU vs US endpoints.

@bernerdschaefer

### CODE REVIEW

File: builtin/providers/heroku/resource_heroku_cert.go [hunk 1]
```diff
@@ -88,22 +88,24 @@ func resourceHerokuCertUpdate(d *schema.ResourceData, meta interface{}) error {
...
+	if d.HasChange("certificate_chain") || d.HasChange("private_key") {
+		log.Printf("[DEBUG] SSL Certificate update configuration: %#v, %#v", app, opts)
+		ad, err := client.SSLEndpointUpdate(context.TODO(), app, d.Id(), opts)
 		if err != nil {
-			return err
+			return fmt.Errorf("Error updating SSL endpoint: %s", err)
 		}
 
 		// Store the new ID
 		d.SetId(ad.ID)
```
Comment (catsby) [inline builtin/providers/heroku/resource_heroku_cert.go]: I know this line isn't changing here, but it caught our eye. Is the ID of the SSL Cert changing? If so, either the `private_key` or `certificate_chain` attribute (or both?) should have `ForceNew` specified and instruct terraform to do a complete destroy and recreate. Updating ID here could result in unexpected behavior (dangling resources), or just undefined behavior, depending on the provider. 

The client methods being called, and the underlying `PATCH` API call, suggests to me the ID is not actually changing and we should remove this line while we are here, if you could.

Comment (wchrisjohnson) [inline builtin/providers/heroku/resource_heroku_cert.go]: Wow, that's great feedback. This is my first PR against Terraform, and I'm still learning how it works. I will do some testing and handle this, or come back with additional questions...

Comment (wchrisjohnson) [inline builtin/providers/heroku/resource_heroku_cert.go]: @catsby I just verified that the ID is NOT changing between the create and the update, so I will remove this line.

**Comment by wchrisjohnson** [comment]
Could someone provide feedback on what, if anything, remaining to be done in order to get this merged?

**Comment by catsby** [comment]
Hey @wchrisjohnson – I ran the tests and got the following error:

```
[pr-14240]$ make testacc TEST=./builtin/providers/heroku TESTARGS="-run=TestAccHerokuCert"
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/05/09 11:20:12 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/heroku -v -run=TestAccHerokuCert -timeout 120m
=== RUN   TestAccHerokuCert_EU
--- FAIL: TestAccHerokuCert_EU (18.65s)
        testing.go:280: Step 0 error: Error applying: 1 error(s) occurred:

                * heroku_cert.ssl_certificate: 1 error(s) occurred:

                * heroku_cert.ssl_certificate: Error creating SSL endpoint: Post https://api.heroku.com/apps/tftest-jtvdruccq7/ssl-endpoints: Conflict
=== RUN   TestAccHerokuCert_US
--- PASS: TestAccHerokuCert_US (32.36s)
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/builtin/providers/heroku 51.030s
make: *** [testacc] Error 1
```

Is that a limitation of my test account?

**Comment by wchrisjohnson** [comment]
@catsby It works on my end. Try the following:
```
$ HEROKU_API_KEY=<REDACTED> make testacc TEST=./builtin/providers/heroku TESTARGS="-run=TestAccHerokuCert"
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/05/09 12:55:29 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/heroku -v -run=TestAccHerokuCert -timeout 120m
=== RUN   TestAccHerokuCert_EU
--- PASS: TestAccHerokuCert_EU (23.20s)
=== RUN   TestAccHerokuCert_US
--- PASS: TestAccHerokuCert_US (20.11s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/heroku	43.333s
```

**Comment by catsby** [comment]
I have all the Heroku things I should need in the environment: 
```
HEROKU_ORGANIZATION=terraform
HEROKU_EMAIL=terraform@...
HEROKU_API_KEY=abc123-redacted
```

I ran the tests again, strangely I'm getting a different error now, this time with the US SSL, not the EU one:

```
[pr-14240](4)$ make testacc TEST=./builtin/providers/heroku TESTARGS="-run=TestAccHerokuCert"
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/05/09 16:20:30 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/heroku -v -run=TestAccHerokuCert -timeout 120m
=== RUN   TestAccHerokuCert_EU
--- PASS: TestAccHerokuCert_EU (22.25s)
=== RUN   TestAccHerokuCert_US
--- FAIL: TestAccHerokuCert_US (13.14s)
        testing.go:280: Step 1 error: Error applying: 1 error(s) occurred:

                * heroku_cert.ssl_certificate: 1 error(s) occurred:

                * heroku_cert.ssl_certificate: Error updating SSL endpoint: Patch https://api.heroku.com/apps/tftest-yf43o3l4ph/ssl-endpoints/9609ba7d-9d82-4426-8cf9-fdbbd9464ec5: Could not complete the action. Please try again later.
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/builtin/providers/heroku 35.422s
make: *** [testacc] Error 1
```

**Comment by catsby** [comment]
I did manage to get them to pass, but the inconsistency is troubling, do you have any thoughts there?

**Comment by wchrisjohnson** [comment]
yeah @catsby I have seen that specific error crop up periodically in my running of the test suite as well. It feels like some sort of race condition, as it doesn't occur consistently. Let me dig into this on my end a bit more.

**Comment by catsby** [comment]
@wchrisjohnson thanks! Feel free to ping me if/when you find out 😄

**Comment by wchrisjohnson** [comment]
@catsby I've been talking to some other engineers here at Heroku; there are some upstream conditions where rapid create/update of addon and SSL endpoints can fail, but can safely be retried. We're diagnosing the issue but believe it's unrelated to the changes in the PR.

**Comment by catsby** [comment]
OK, sounds good 👍

**Pull request merged**